rotkehlchen/chain/ethereum/modules/eth2/decoder.py

```diff
@@ -0,0 +1,79 @@
+from rotkehlchen.accounting.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.chain.ethereum.constants import ETH2_DEPOSIT_ADDRESS
+from rotkehlchen.chain.evm.decoding.interfaces import DecoderInterface
+from rotkehlchen.chain.evm.decoding.structures import ActionItem
+from rotkehlchen.chain.evm.structures import EvmTxReceiptLog
+from rotkehlchen.constants.assets import A_ETH
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import ChecksumEvmAddress, EvmTransaction
+from rotkehlchen.utils.misc import from_gwei, hex_or_bytes_to_int
+
+from .constants import CPT_ETH2
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.ethereum.node_inquirer import EthereumInquirer
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.user_messages import MessagesAggregator
+
+
+DEPOSIT_EVENT = b'd\x9b\xbcb\xd0\xe3\x13B\xaf\xeaN\\\xd8-@I\xe7\xe1\xee\x91/\xc0\x88\x9a\xa7\x90\x80;\xe3\x908\xc5'  # noqa: E501
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class Eth2Decoder(DecoderInterface):
+    def __init__(  # pylint: disable=super-init-not-called
+            self,
+            ethereum_inquirer: 'EthereumInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',  # pylint: disable=unused-argument
+    ) -> None:
+        self.base = base_tools
+        self.ethereum = ethereum_inquirer
+
+    def _decode_eth2_deposit_event(
+            self,
+            tx_log: EvmTxReceiptLog,
+            transaction: EvmTransaction,  # pylint: disable=unused-argument
+            decoded_events: list[HistoryBaseEntry],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: Optional[list[ActionItem]],  # pylint: disable=unused-argument
+    ) -> tuple[Optional[HistoryBaseEntry], list[ActionItem]]:
+        if tx_log.topics[0] != DEPOSIT_EVENT:
+            return None, []
+
+        pubkey = encode_hex(tx_log.data[192:240])
+        withdrawal_credentials = encode_hex(tx_log.data[288:320])
+        amount = from_gwei(hex_or_bytes_to_int(tx_log.data[352:360], byteorder='little'))
+        deposit_index = hex_or_bytes_to_int(tx_log.data[544:552 + 8], byteorder='little')  # is not same as validator index  # noqa: E501
+        for event in decoded_events:
+            if (
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.asset == A_ETH and
+                event.balance.amount == amount
+            ):
+                event.event_type = HistoryEventType.STAKING
+                event.event_subtype = HistoryEventSubType.DEPOSIT_ASSET
+                event.counterparty = CPT_ETH2
+                event.notes = f'Deposit {amount} ETH to validator with pubkey {pubkey}. Deposit index: {deposit_index}'  # noqa: E501
```

Comment:
Having withdrawal credentials in the notes is not bad either since this is what is shown to the user.